### PR TITLE
feat(memft): token-weighted memorization loss with fused linear-CE kernel

### PR DESCRIPTION
**Pull request by winglian**
Implements **MemFT** (Memorization-oriented Fine-Tuning) from *"How LoRA Remembers?
A Parametric Memory Law for LLM Finetuning"* ([arXiv:2605.30260](https://arxiv.org/abs/2605.30260)).

## Variants

Configured under a nested `memft:` object:

- **`ot`** (only-threshold): hard mask `w_t = 1[L_t > L_crit]`. No extra hyper-parameters.
- **`sw`** (sliding-window): soft threshold `σ(κ(L_t − L_crit))` modulated by an
  exponential spatial decay anchored at the first greedy-prediction error.

```yaml
memft:
  variant: ot
  fused: true                 # variant 'ot' only — fused linear-CE Triton kernel
  # fused_chunk_tokens: 2048  # tune the speed/memory frontier
  # critical_loss: 0.6931     # defaults to ln(2)
  # sw: { kappa: 1.0, tau: 64.0, window: 64, floor: 0.0 }
```

Fused linear cross-entropy kernel (OT)

The fused: true path computes the loss directly from hidden states and the LM-head
weight via a Triton kernel + chunked autograd Function, never materializing the
[tokens, vocab] logits. Because the OT weights are detached, the weight-sum
normalizer is a scalar w.r.t. autograd, so a single chunked pass is exact (gradients
accumulate at hidden/weight size and are scaled once at the end).

Memory (Llama-3 vocab, H=4096, bf16, fwd+bwd, RTX PRO 6000):
```yaml
┌────────┬───────────┬────────┬───────────┐
│ tokens │ non-fused │ fused  │ reduction │
├────────┼───────────┼────────┼───────────┤
│   4096 │    9.5 GB │ 8.5 GB │       11% │
├────────┼───────────┼────────┼───────────┤
│   8192 │   17.0 GB │ 7.6 GB │       49% │
├────────┼───────────┼────────┼───────────┤
│  16384 │   37.7 GB │ 9.8 GB │       74% │
└────────┴───────────┴────────┴───────────┘
```

Packing-aware

The loss is computed inside a patched llama-like CausalLM.forward, so it can see
position_ids. Under sample_packing, the label shift and the MemFT-SW anchor/window
are confined to each individual sample (no cross-sample leakage). The non-fused path
still returns logits so eval metrics keep working.

Correctness

- Fused matches an fp32 reference exactly in fp32; bf16 grad rel-error ~1.5%, bounded by
fp32 grad-weight accumulation (regression-tested under many chunks).
- Fused-packed == non-fused-packed (diff 0.0).
- The Triton kernel + autograd math were reviewed by a kernel-correctness pass; findings
(bf16 accumulator drift, OOB label load) were fixed and covered by tests.

Notes / limitations

- MemFT is mutually exclusive with the other cross-entropy optimizations
(CCE, chunked CE, Liger CE/FLCE), enforced in config validation.
- Computing the loss in the forward limits MemFT to llama-like architectures
(llama, qwen, mistral, …); other model types raise a clear error.
- fused: true supports variant: ot only (rejected for sw in config validation).

## Release Notes

* **New Features**
  * Added MemFT (Memorization-oriented Fine-Tuning) training capability with two variants (ot and sw).
  * Added configuration support and example training setup for MemFT with LoRA adapters.

* **Documentation**
  * Added comprehensive MemFT documentation with configuration guidance and variants description.

* **Tests**
  * Added test coverage for MemFT functionality.

### CODE REVIEW

**Review by ved1beta — COMMENTED** [review]
overall everything looks good small concerns, 
was skimming through the paper, 
noticed the sliding window is fixed i according to the paper it should increase with proportionally to a_i
